File: installer/nim/installer.nim
# NullKia Installer - Nim
# Compiles to native C for stealth and performance
# @author @AnonAntics
# @discord discord.gg/killers

import os, strutils, terminal, times, httpclient, json, osproc
import std/[asyncdispatch, asynchttpserver, md5]

const
  VERSION = "2.0.0"
  BANNER = """
  ╭──────────────────────────────────────────╮
  │           📱 NULLKIA INSTALLER           │
  │        ════════════════════════          │
  │                                          │
  │   🔧 Installation Wizard v2.0.0          │
  │   📡 Stealth Mode: ENABLED               │
  │   💾 Target: System Integration          │
  │                                          │
  │        Press ENTER to continue...        │
  ╰──────────────────────────────────────────╯
              @AnonAntics | NullSec
  """
  
  LICENSE_PREFIX = "NKIA"
  DISCORD_URL = "https://discord.gg/killers"
  INSTALL_PATHS = [
    "/opt/nullkia",
    "/usr/local/nullkia",
    "~/.nullkia"
  ]

type
  InstallMode = enum
    imFull, imMinimal, imPortable, imCustom
  
  LicenseTier = enum
    ltFree, ltPremium, ltEnterprise
  
  InstallConfig = object
    mode: InstallMode
    path: string
    license: LicenseTier
    key: string
    components: seq[string]
    createDesktop: bool
    addToPath: bool
    startOnBoot: bool

  InstallerState = object
    config: InstallConfig
    progress: int
    currentStep: string
    errors: seq[string]

# Terminal colors
proc setColor(fg: ForegroundColor) =
  stdout.setForegroundColor(fg)

proc resetColor() =
  stdout.resetAttributes()

proc printBanner() =
  setColor(fgCyan)
  echo BANNER
  resetColor()

proc printStep(step: string, status: string = "...") =
  setColor(fgYellow)
  stdout.write("[")
  setColor(fgCyan)
  stdout.write(step)
  setColor(fgYellow)
  stdout.write("] ")
  resetColor()
  echo status

proc printSuccess(msg: string) =
  setColor(fgGreen)
  echo "✅ " & msg
  resetColor()

proc printError(msg: string) =
  setColor(fgRed)
  echo "❌ " & msg
  resetColor()

proc printWarning(msg: string) =
  setColor(fgYellow)
  echo "⚠️  " & msg
  resetColor()

# License validation
proc validateLicense(key: string): tuple[valid: bool, tier: LicenseTier, msg: string] =
  # Format: NKIA-XXXX-XXXX-XXXX-XXXX
  if key.len != 24:
    return (false, ltFree, "Invalid key length")
  
  if not key.startsWith(LICENSE_PREFIX):
    return (false, ltFree, "Invalid key prefix")
  
  let parts = key.split("-")
  if parts.len != 5:
    return (false, ltFree, "Invalid key format")
  
  # Determine tier from second segment
  let tierCode = parts[1][0..1]
  var tier = ltFree
  
  case tierCode:
    of "PR": tier = ltPremium
    of "EN": tier = ltEnterprise
    of "FR": tier = ltFree
    else: tier = ltPremium
  
  # Simple checksum validation
  var checksum = 0
  for c in key:
    checksum = checksum xor ord(c)
  
  return (true, tier, "License valid: " & $tier)

# Component detection
proc detectSystem(): tuple[os: string, arch: string, isRoot: bool] =
  let osName = when defined(linux): "linux"
               elif defined(macosx): "macos"
               elif defined(windows): "windows"
               else: "unknown"
  
  let arch = when defined(amd64): "x64"
             elif defined(arm64): "arm64"
             elif defined(i386): "x86"
             else: "unknown"
  
  let isRoot = when defined(windows): true
               else: getEnv("USER") == "root"
  
  return (osName, arch, isRoot)

# Installation components
let ALL_COMPONENTS = @[
  "core",           # Core framework
  "firmware",       # Firmware tools
  "baseband",       # Baseband utilities
  "samsung",        # Samsung specific
  "apple",          # Apple specific
  "xiaomi",         # Xiaomi specific
  "google",         # Google/Pixel specific
  "oneplus",        # OnePlus specific
  "huawei",         # Huawei specific
  "motorola",       # Motorola specific
  "forensics",      # Forensic tools
  "unlock",         # Unlock utilities
  "gui"             # GUI components
]

proc selectComponents(): seq[string] =
  echo "\n📦 Select components to install:\n"
  
  for i, comp in ALL_COMPONENTS:
    echo "  [" & $(i+1) & "] " & comp
  
  echo "\n  [A] All components"
  echo "  [M] Minimal (core only)"
  echo "  [C] Custom selection"
  
  stdout.write("\nSelection: ")
  let choice = stdin.readLine().toLowerAscii()
  
  case choice:
    of "a", "all":
      return ALL_COMPONENTS
    of "m", "minimal":
      return @["core"]
    of "c", "custom":
      stdout.write("Enter component numbers (comma-separated): ")
      let nums = stdin.readLine().split(",")
      var selected: seq[string] = @[]
      for num in nums:
        let idx = parseInt(num.strip()) - 1
        if idx >= 0 and idx < ALL_COMPONENTS.len:
          selected.add(ALL_COMPONENTS[idx])
      return selected
    else:
      return @["core", "firmware", "baseband"]

proc selectInstallPath(): string =
  echo "\n📁 Select installation path:\n"
  
  for i, path in INSTALL_PATHS:
    echo "  [" & $(i+1) & "] " & path
  
  echo "  [C] Custom path"
  
  stdout.write("\nSelection: ")
  let choice = stdin.readLine()
  
  try:
    let idx = parseInt(choice) - 1
    if idx >= 0 and idx < INSTALL_PATHS.len:
      return INSTALL_PATHS[idx].replace("~", getHomeDir())
  except:
    discard
  
  if choice.toLowerAscii() == "c":
    stdout.write("Enter custom path: ")
    return stdin.readLine()
  
  return INSTALL_PATHS[0]

proc createDirectories(basePath: string, components: seq[string]) =
  printStep("Creating directories")
  
  createDir(basePath)
  createDir(basePath / "bin")
  createDir(basePath / "lib")
  createDir(basePath / "config")
  createDir(basePath / "data")
  createDir(basePath / "logs")
  createDir(basePath / "cache")
  
  for comp in components:
    createDir(basePath / "modules" / comp)
  
  printSuccess("Directory structure created")

proc writeConfig(basePath: string, config: InstallConfig) =
  printStep("Writing configuration")
  
  let configJson = %* {
    "version": VERSION,
    "install_date": $now(),
    "install_path": basePath,
    "license_tier": $config.license,
    "components": config.components,
    "author": "@AnonAntics",
    "discord": DISCORD_URL
  }
  
  writeFile(basePath / "config" / "nullkia.json", $configJson)
  printSuccess("Configuration saved")

proc installCore(basePath: string) =
  printStep("Installing core framework")
  
  # Create main launcher script
  let launcher = """#!/bin/bash
# NullKia Launcher
# @AnonAntics | discord.gg/killers

NULLKIA_HOME="$BASEPATH"
export NULLKIA_HOME

exec "$NULLKIA_HOME/bin/nullkia-core" "$@"
""".replace("$BASEPATH", basePath)
  
  writeFile(basePath / "bin" / "nullkia", launcher)
  
  when not defined(windows):
    discard execCmd("chmod +x " & basePath / "bin" / "nullkia")
  
  printSuccess("Core framework installed")

proc createDesktopEntry(basePath: string) =
  printStep("Creating desktop entry")
  
  when defined(linux):
    let desktopEntry = """[Desktop Entry]
Version=1.0
Type=Application
Name=NullKia
Comment=Mobile Security Framework
Exec=$BASEPATH/bin/nullkia
Icon=$BASEPATH/icons/nullkia.png
Terminal=true
Categories=Security;Development;
Keywords=mobile;security;firmware;hacking;
""".replace("$BASEPATH", basePath)
    
    let desktopPath = getHomeDir() / ".local" / "share" / "applications"
    createDir(desktopPath)
    writeFile(desktopPath / "nullkia.desktop", desktopEntry)
    printSuccess("Desktop entry created")
  
  elif defined(macosx):
    printWarning("macOS: Use Applications folder manually")
  
  elif defined(windows):
    printWarning("Windows: Shortcut creation requires admin")

proc addToPath(basePath: string) =
  printStep("Adding to PATH")
  
  when defined(linux) or defined(macosx):
    let shellRc = if fileExists(getHomeDir() / ".zshrc"):
                    getHomeDir() / ".zshrc"
                  else:
                    getHomeDir() / ".bashrc"
    
    let pathLine = "\n# NullKia\nexport PATH=\"$PATH:" & basePath & "/bin\"\n"
    
    var content = readFile(shellRc)
    if not content.contains("NullKia"):
      appendFile(shellRc, pathLine)
      printSuccess("Added to PATH in " & shellRc)
    else:
      printWarning("Already in PATH")
  
  elif defined(windows):
    printWarning("Windows: Add to PATH manually")

proc runInstaller() =
  printBanner()
  discard stdin.readLine()
  
  let (osName, arch, isRoot) = detectSystem()
  
  echo "\n🔍 System Detection:"
  echo "   OS: " & osName
  echo "   Arch: " & arch
  echo "   Root: " & $isRoot
  
  if not isRoot:
    printWarning("Running without root. Some features may be limited.")
  
  # License input
  echo "\n🔑 License Activation"
  echo "   Get your key at: " & DISCORD_URL
  stdout.write("   Enter license key (or press ENTER for free): ")
  let licenseKey = stdin.readLine()
  
  var config = InstallConfig(
    mode: imFull,
    license: ltFree,
    key: "",
    components: @[],
    createDesktop: true,
    addToPath: true,
    startOnBoot: false
  )
  
  if licenseKey.len > 0:
    let (valid, tier, msg) = validateLicense(licenseKey)
    if valid:
      printSuccess(msg)
      config.license = tier
      config.key = licenseKey
    else:
      printError(msg)
      config.license = ltFree
  else:
    printWarning("Using free tier. Premium at discord.gg/killers")
  
  # Select components
  config.components = selectComponents()
  echo "\n📦 Components: " & $config.components.len & " selected"
  
  # Select path
  config.path = selectInstallPath()
  echo "📁 Install path: " & config.path
  
  # Confirm
  echo "\n" & "─".repeat(50)
  echo "📋 Installation Summary:"
  echo "   Path: " & config.path
  echo "   License: " & $config.license
  echo "   Components: " & config.components.join(", ")
  echo "─".repeat(50)
  
  stdout.write("\nProceed with installation? [Y/n]: ")
  let confirm = stdin.readLine().toLowerAscii()
  
  if confirm != "" and confirm != "y" and confirm != "yes":
    printError("Installation cancelled")
    return
  
  echo "\n🚀 Installing NullKia...\n"
  
  # Run installation steps
  createDirectories(config.path, config.components)
  writeConfig(config.path, config)
  installCore(config.path)
  
  if config.createDesktop:
    createDesktopEntry(config.path)
  
  if config.addToPath:
    addToPath(config.path)
  
  echo "\n" & "═".repeat(50)
  setColor(fgGreen)
  echo "✅ NullKia installed successfully!"
  resetColor()
  echo "═".repeat(50)
  echo "\n   Run 'nullkia' to start"
  echo "   Documentation: " & config.path & "/docs"
  echo "   Discord: " & DISCORD_URL
  echo "   Twitter: @AnonAntics"
  echo ""

when isMainModule:
  runInstaller()

File: installer/zig/installer.zig
// ═══════════════════════════════════════════════════════════════════
//  NULLKIA ZIG INSTALLER
//  High-performance system installer in Zig
//  @author bad-antics | discord.gg/killers
// ═══════════════════════════════════════════════════════════════════

const std = @import("std");
const fs = std.fs;
const mem = std.mem;
const os = std.os;
const io = std.io;

const VERSION = "2.0.0";
const AUTHOR = "bad-antics";
const DISCORD = "discord.gg/killers";

const BANNER =
    \\
    \\ ╭──────────────────────────────────────────╮
    \\ │       📱 NULLKIA ZIG INSTALLER           │
    \\ │       ════════════════════════           │
    \\ │                                          │
    \\ │   ⚡ Zero-Cost Abstractions              │
    \\ │   🔧 Memory-Safe Installation            │
    \\ │   💾 Compile-Time Verification           │
    \\ │                                          │
    \\ │          bad-antics | NullSec            │
    \\ ╰──────────────────────────────────────────╯
    \\
;

// ═══════════════════════════════════════════════════════════════════
// License Management
// ═══════════════════════════════════════════════════════════════════

const LicenseTier = enum {
    Free,
    Premium,
    Enterprise,
};

const License = struct {
    key: []const u8,
    tier: LicenseTier,
    valid: bool,
};

fn validateLicense(key: []const u8) License {
    if (key.len != 24) {
        return License{ .key = "", .tier = .Free, .valid = false };
    }

    if (!mem.startsWith(u8, key, "NKIA-")) {
        return License{ .key = "", .tier = .Free, .valid = false };
    }

    // Check tier based on key pattern
    var tier: LicenseTier = .Free;
    if (key.len > 5) {
        if (mem.startsWith(u8, key[5..], "PR")) {
            tier = .Premium;
        } else if (mem.startsWith(u8, key[5..], "EN")) {
            tier = .Enterprise;
        }
    }

    return License{
        .key = key,
        .tier = tier,
        .valid = true,
    };
}

// ═══════════════════════════════════════════════════════════════════
// Installation Components
// ═══════════════════════════════════════════════════════════════════

const Component = struct {
    name: []const u8,
    description: []const u8,
    size_mb: u32,
    requires_premium: bool,
    selected: bool,
};

const components = [_]Component{
    .{ .name = "core", .description = "NullKia Core Framework", .size_mb = 50, .requires_premium = false, .selected = true },
    .{ .name = "samsung", .description = "Samsung/Knox Tools", .size_mb = 120, .requires_premium = false, .selected = true },
    .{ .name = "apple", .description = "Apple/iOS Tools", .size_mb = 150, .requires_premium = true, .selected = false },
    .{ .name = "google", .description = "Google Pixel/Titan M", .size_mb = 100, .requires_premium = false, .selected = true },
    .{ .name = "oneplus", .description = "OnePlus Tools", .size_mb = 80, .requires_premium = false, .selected = false },
    .{ .name = "xiaomi", .description = "Xiaomi/MIUI Tools", .size_mb = 90, .requires_premium = false, .selected = false },
    .{ .name = "huawei", .description = "Huawei/EMUI Tools", .size_mb = 85, .requires_premium = true, .selected = false },
    .{ .name = "firmware", .description = "Firmware Analysis Suite", .size_mb = 200, .requires_premium = true, .selected = false },
    .{ .name = "baseband", .description = "Baseband Research Tools", .size_mb = 180, .requires_premium = true, .selected = false },
    .{ .name = "exploits", .description = "Exploit Database", .size_mb = 300, .requires_premium = true, .selected = false },
};

// ═══════════════════════════════════════════════════════════════════
// Console Output
// ═══════════════════════════════════════════════════════════════════

fn printColored(comptime color: []const u8, msg: []const u8) void {
    const stdout = std.io.getStdOut().writer();
    stdout.print("\x1b[{s}m{s}\x1b[0m\n", .{ color, msg }) catch {};
}

fn printSuccess(msg: []const u8) void {
    printColored("32", msg);
}

fn printError(msg: []const u8) void {
    printColored("31", msg);
}

fn printWarning(msg: []const u8) void {
    printColored("33", msg);
}

fn printInfo(msg: []const u8) void {
    printColored("36", msg);
}

// ═══════════════════════════════════════════════════════════════════
// Progress Bar
// ═══════════════════════════════════════════════════════════════════

fn drawProgressBar(current: u32, total: u32, width: u32) void {
    const stdout = std.io.getStdOut().writer();
    const progress = @as(f32, @floatFromInt(current)) / @as(f32, @floatFromInt(total));
    const filled = @as(u32, @intFromFloat(progress * @as(f32, @floatFromInt(width))));
    
    stdout.print("\r  [", .{}) catch {};
    
    var i: u32 = 0;
    while (i < width) : (i += 1) {
        if (i < filled) {
            stdout.print("█", .{}) catch {};
        } else {
            stdout.print("░", .{}) catch {};
        }
    }
    
    const percent = @as(u32, @intFromFloat(progress * 100));
    stdout.print("] {d}%", .{percent}) catch {};
}

// ═══════════════════════════════════════════════════════════════════
// Installation Functions
// ═══════════════════════════════════════════════════════════════════

fn createDirectory(path: []const u8) !void {
    fs.cwd().makePath(path) catch |err| {
        if (err != error.PathAlreadyExists) {
            return err;
        }
    };
}

fn installComponent(comp: Component, install_path: []const u8, allocator: std.mem.Allocator) !void {
    const stdout = std.io.getStdOut().writer();
    
    stdout.print("\n  📦 Installing {s}...\n", .{comp.name}) catch {};
    stdout.print("     {s}\n", .{comp.description}) catch {};
    
    // Create component directory
    var path_buf: [512]u8 = undefined;
    const comp_path = std.fmt.bufPrint(&path_buf, "{s}/{s}", .{ install_path, comp.name }) catch return;
    try createDirectory(comp_path);
    
    // Simulate installation progress
    const steps: u32 = 20;
    var i: u32 = 0;
    while (i <= steps) : (i += 1) {
        drawProgressBar(i, steps, 30);
        std.time.sleep(50_000_000); // 50ms
    }
    
    stdout.print("\n     ✅ Installed: {d} MB\n", .{comp.size_mb}) catch {};
    _ = allocator;
}

fn performInstallation(license: License, install_path: []const u8, allocator: std.mem.Allocator) !void {
    const stdout = std.io.getStdOut().writer();
    
    stdout.print("\n═══════════════════════════════════════════\n", .{}) catch {};
    stdout.print("  📱 NULLKIA INSTALLATION\n", .{}) catch {};
    stdout.print("═══════════════════════════════════════════\n", .{}) catch {};
    
    stdout.print("\n  📂 Install Path: {s}\n", .{install_path}) catch {};
    stdout.print("  🔑 License: {s}\n", .{if (license.valid) "Valid" else "Free"}) catch {};
    
    // Create base directory
    try createDirectory(install_path);
    
    // Install selected components
    var total_size: u32 = 0;
    var installed: u32 = 0;
    
    for (components) |comp| {
        if (comp.selected) {
            if (comp.requires_premium and license.tier == .Free) {
                stdout.print("\n  ⚠️  Skipping {s} (Premium required)\n", .{comp.name}) catch {};
                stdout.print("     🔑 Get premium at discord.gg/killers\n", .{}) catch {};
                continue;
            }
            
            try installComponent(comp, install_path, allocator);
            total_size += comp.size_mb;
            installed += 1;
        }
    }
    
    stdout.print("\n═══════════════════════════════════════════\n", .{}) catch {};
    stdout.print("  ✅ INSTALLATION COMPLETE\n", .{}) catch {};
    stdout.print("═══════════════════════════════════════════\n", .{}) catch {};
    stdout.print("\n  📦 Components: {d}\n", .{installed}) catch {};
    stdout.print("  💾 Total Size: {d} MB\n", .{total_size}) catch {};
    stdout.print("  📂 Location: {s}\n", .{install_path}) catch {};
}

// ═══════════════════════════════════════════════════════════════════
// System Detection
// ═══════════════════════════════════════════════════════════════════

const SystemInfo = struct {
    os_name: []const u8,
    arch: []const u8,
    home_dir: []const u8,
};

fn detectSystem() SystemInfo {
    const os_tag = @import("builtin").os.tag;
    const arch_tag = @import("builtin").cpu.arch;
    
    const os_name = switch (os_tag) {
        .linux => "Linux",
        .macos => "macOS",
        .windows => "Windows",
        else => "Unknown",
    };
    
    const arch = switch (arch_tag) {
        .x86_64 => "x86_64",
        .aarch64 => "arm64",
        .arm => "arm",
        else => "unknown",
    };
    
    const home = std.os.getenv("HOME") orelse "/tmp";
    
    return SystemInfo{
        .os_name = os_name,
        .arch = arch,
        .home_dir = home,
    };
}

// ═══════════════════════════════════════════════════════════════════
// Main Entry Point
// ═══════════════════════════════════════════════════════════════════

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    
    const stdout = std.io.getStdOut().writer();
    const stdin = std.io.getStdIn().reader();
    
    // Print banner
    stdout.print("{s}\n", .{BANNER}) catch {};
    stdout.print("  Version {s} | {s}\n", .{ VERSION, AUTHOR }) catch {};
    stdout.print("  🔑 Premium: {s}\n\n", .{DISCORD}) catch {};
    
    // Detect system
    const sys = detectSystem();
    stdout.print("  💻 System: {s} ({s})\n", .{ sys.os_name, sys.arch }) catch {};
    
    // Get license key
    stdout.print("\n  Enter license key (or press Enter for free): ", .{}) catch {};
    
    var key_buf: [256]u8 = undefined;
    const key_len = stdin.readUntilDelimiter(&key_buf, '\n') catch |err| {
        if (err == error.EndOfStream) {
            return;
        }
        return err;
    };
    
    const key = mem.trim(u8, key_buf[0..key_len.len], &[_]u8{ '\r', '\n', ' ' });
    const license = validateLicense(key);
    
    if (license.valid) {
        printSuccess("  ✅ License validated!");
        stdout.print("     Tier: {s}\n", .{@tagName(license.tier)}) catch {};
    } else if (key.len > 0) {
        printWarning("  ⚠️  Invalid license, using free tier");
    }
    
    // Default install path
    var path_buf: [512]u8 = undefined;
    const default_path = std.fmt.bufPrint(&path_buf, "{s}/.nullkia", .{sys.home_dir}) catch "/tmp/.nullkia";
    
    // Confirm installation
    stdout.print("\n  Install to {s}? [Y/n]: ", .{default_path}) catch {};
    
    var confirm_buf: [16]u8 = undefined;
    const confirm_len = stdin.readUntilDelimiter(&confirm_buf, '\n') catch "";
    const confirm = mem.trim(u8, confirm_buf[0..confirm_len.len], &[_]u8{ '\r', '\n', ' ' });
    
    if (confirm.len > 0 and (confirm[0] == 'n' or confirm[0] == 'N')) {
        printInfo("  Installation cancelled.");
        return;
    }
    
    // Perform installation
    try performInstallation(license, default_path, allocator);
    
    // Print footer
    stdout.print("\n─────────────────────────────────────────\n", .{}) catch {};
    stdout.print("  📱 NullKia Zig Installer\n", .{}) catch {};
    stdout.print("  🔑 Premium: discord.gg/killers\n", .{}) catch {};
    stdout.print("  👤 Author: bad-antics\n", .{}) catch {};
    stdout.print("─────────────────────────────────────────\n\n", .{}) catch {};
}
